libr/util/event.c

```diff
@@ -34,50 +43,64 @@ R_API void r_event_free(REvent *ev) {
 	free (ev);
 }
 
-static bool add_hook(void *cb, const ut64 k, const void *v) {
-	RPVector *cbs = (RPVector *)v;
+static bool add_hook(void *hook, const ut64 k, const void *v) {
+	RVector *cbs = (RVector *)v;
 	r_return_val_if_fail (cbs, false);
-	r_pvector_push (cbs, cb);
+	r_vector_push (cbs, hook);
 	return true;
 }
 
-R_API void r_event_hook(REvent *ev, int type, REventCallback cb) {
-	r_return_if_fail (ev);
+R_API REventCallbackHandle r_event_hook(REvent *ev, int type, REventCallback cb, void *user) {
+	REventCallbackHandle handle = { 0 };
+	r_return_val_if_fail (ev, handle);
+	REventCallbackHook hook;
+	hook.cb = cb;
+	hook.user = user;
+	hook.handle = ev->hook_handle_next++;
 	if (type == R_EVENT_ALL) {
-		ht_up_foreach (ev->callbacks, add_hook, cb);
+		ht_up_foreach (ev->callbacks, add_hook, &hook);
 	} else {
-		RPVector *cbs = ht_up_find (ev->callbacks, type, NULL);
-		add_hook (cb, 0, cbs);
+		RPVector *cbs = ht_up_find (ev->callbacks, (ut64)type, NULL);
+		add_hook (&hook, 0, cbs);
 	}
+	handle.handle = hook.handle;
+	handle.type = type;
+	return handle;
 }
 
-static bool del_hook(void *cb, const ut64 k, const void *v) {
-	RPVector *cbs = (RPVector *)v;
+static bool del_hook(void *user, const ut64 k, const void *v) {
+	int handle = *((int *)user);
+	RVector *cbs = (RVector *)v;
 	r_return_val_if_fail (cbs, false);
-	r_pvector_remove_data (cbs, cb);
+	size_t i;
+	for (i = 0; i < cbs->len; i++) {
```

Comment:
what about a `r_vector_enumerate(vector, it, i)`? it would be a bit more declarative.